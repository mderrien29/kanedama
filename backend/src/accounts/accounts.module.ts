import { HttpModule, Module } from '@nestjs/common';

import { AccountsService } from './accounts.service';
import { ApiService } from './api.service';
import { MetricsService } from './metrics.service';

@Module({
  imports: [HttpModule],
  providers: [AccountsService, ApiService, MetricsService],
  exports: [AccountsService, MetricsService],
})
export class AccountsModule {}
