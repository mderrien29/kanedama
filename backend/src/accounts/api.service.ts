import { Injectable, HttpService } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { map } from 'rxjs/operators';

import { TransactionDto } from 'src/common/dtos/transaction.dto';
import { AccountDto } from 'src/common/dtos/account.dto';

@Injectable()
export class ApiService {
  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {}

  private get accountsBaseUrl() {
    return this.configService.get('ACCOUNT_URI');
  }

  private async request<T>(url: string): Promise<T> {
    return await this.httpService
      .get(url)
      .pipe(map((response) => response.data))
      .toPromise();
  }

  public async getAccounts(): Promise<AccountDto[]> {
    const url = this.accountsBaseUrl;
    const response = await this.request(url);

    if (!this.isAccountsDto(response)) {
      throw new Error('retrieved payload does not match the expected format');
    }

    return response;
  }

  private isAccountsDto(payload: any): payload is AccountDto[] {
    // TODO
    return true;
  }

  public async getOldestTransaction(account: string): Promise<TransactionDto> {
    const url = `${this.accountsBaseUrl}/${account}/transactions`;
    const response = await this.request(url);

    if (!this.isTransactionDto(response)) {
      throw new Error('retrieved payload does not match the expected format');
    }

    return response;
  }

  private isTransactionDto(payload: any): payload is TransactionDto {
    // TODO
    return true;
  }

  public async getTransactions(
    account: string,
    startDate?: Date,
    endDate?: Date,
  ): Promise<TransactionDto[]> {
    const transactionsUrl = this.formatTransactionUrl(
      account,
      startDate,
      endDate,
    );
    const response = await this.request(transactionsUrl);

    if (!this.isTransactionDtoArray(response)) {
      throw new Error('retrieved payload does not match the expected format');
    }

    return response;
  }

  // TODO use package to add params
  private formatTransactionUrl(
    account: string,
    startDate: Date,
    endDate: Date,
  ): string {
    return `${
      this.accountsBaseUrl
    }/${account}/transactions?from=${startDate.toISOString()}&to=${endDate.toISOString()}`;
  }

  private isTransactionDtoArray(payload: any): payload is TransactionDto[] {
    return payload.every(this.isTransactionDto.bind(this));
  }
}
