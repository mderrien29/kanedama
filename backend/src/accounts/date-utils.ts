import * as chunkDateRange from 'chunk-date-range';

export interface DateInterval {
  start: Date;
  end: Date;
}

export const substractYearsToDate = (
  numberOfYears: number,
  startDate = new Date(),
): Date => {
  return new Date(
    startDate.setFullYear(startDate.getFullYear() - numberOfYears),
  );
};

export const substractMonthToDate = (
  numberOfMonths: number,
  startDate = new Date(),
): Date => {
  return new Date(startDate.setMonth(startDate.getMonth() - numberOfMonths));
};

// Technically, not yearly, but in chunks of 365 days
export const getYearlyIntervalFromDate = (
  startDate: Date,
  endDate = new Date(),
): DateInterval[] => {
  const millisecondsIn365days = 365 * 24 * 60 * 60 * 1000;
  const differenceInMilliseconds = endDate.getTime() - startDate.getTime();
  const numberOfChunks = differenceInMilliseconds / millisecondsIn365days;

  const dateIntervals = chunkDateRange(startDate, endDate, numberOfChunks);
  return dateIntervals;
};
