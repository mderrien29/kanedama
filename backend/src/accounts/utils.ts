export const getDateYearsBefore = (
  numberOfYears: number,
  startDate = new Date(),
): Date => {
  return new Date(
    startDate.setFullYear(startDate.getFullYear() - numberOfYears),
  );
};

export const getDateMonthBefore = (
  numberOfMonths: number,
  startDate = new Date(),
): Date => {
  return new Date(startDate.setMonth(startDate.getMonth() - numberOfMonths));
};
