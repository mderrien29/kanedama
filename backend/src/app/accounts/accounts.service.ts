import { Injectable } from '@nestjs/common';
import * as chunkDateRange from 'chunk-date-range';

import { ApiService } from './api.service';
import { TransactionDto } from 'src/common/dtos/transaction.dto';
import { DateInterval } from 'src/common/interfaces';
import { AccountDto } from 'src/common/dtos/account.dto';

@Injectable()
export class AccountsService {
  constructor(private readonly apiService: ApiService) {}

  public async getUserAccounts(): Promise<AccountDto[]> {
    return await this.apiService.getAccounts();
  }

  public async getUserFullHistory(
    accounts: AccountDto[],
  ): Promise<TransactionDto[]> {
    const allTransactionsFromAllAccounts = accounts.map(
      async (account) =>
        await this.getAllTransactionsFromAccount(account.account_id),
    );

    const transactionsOrderedByDate = (
      await Promise.all(allTransactionsFromAllAccounts)
    )
      .flat(1)
      .sort(this.compareTransactionDate);

    return transactionsOrderedByDate;
  }

  private compareTransactionDate(a: TransactionDto, b: TransactionDto): number {
    return new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime();
  }

  private async getAllTransactionsFromAccount(
    accountId: string,
  ): Promise<TransactionDto[]> {
    const oldestTransaction = await this.apiService.getOldestTransaction(
      accountId,
    );
    const dateIntervals = this.getYearlyIntervalFromDate(
      new Date(oldestTransaction?.timestamp),
    );

    const transactionsPerDateInterval: Promise<TransactionDto[]>[] =
      dateIntervals.map(async (interval: DateInterval) =>
        this.apiService.getTransactions(
          accountId,
          interval.start,
          interval.end,
        ),
      );

    const accountTransactions = (
      await Promise.all(transactionsPerDateInterval)
    ).flat(1);

    return accountTransactions;
  }

  // The api may reject the values due to gap years ? to investigate
  private getYearlyIntervalFromDate(
    startDate: Date,
    endDate = new Date(),
  ): DateInterval[] {
    const numberOfChunks = endDate.getFullYear() - startDate.getFullYear();
    const dateIntervals = chunkDateRange(startDate, endDate, numberOfChunks);
    return dateIntervals;
  }

  // Note: this is valid for amounts in a single currency
  public calculateAccountAmountOverTimeFromCurrent(
    transactions: TransactionDto[],
    currentAccountValue = 0,
  ): number[] {
    const amountOverTime = [currentAccountValue];

    for (let i = 0; i < transactions.length; i++) {
      amountOverTime[i + 1] =
        amountOverTime[i] - transactions[transactions.length - i - 1].amount;
    }

    // TODO could optimise, and move to separate function ofc
    console.dir(Math.floor(Math.max(...amountOverTime)));
    console.dir(Math.floor(Math.min(...amountOverTime)));

    return amountOverTime;
  }

  public sumAccountsCurrentAmount(accounts: AccountDto[]): number {
    return accounts.reduce((sum, account) => (sum += account.current), 0);
  }
}
