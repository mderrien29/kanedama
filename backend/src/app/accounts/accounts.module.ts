import { HttpModule, Module } from '@nestjs/common';

import { AccountsService } from './accounts.service';

@Module({
  imports: [HttpModule],
  providers: [AccountsService],
  exports: [AccountsService],
})
export class AccountsModule {}
