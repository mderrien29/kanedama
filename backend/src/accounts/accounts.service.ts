import { Injectable } from '@nestjs/common';

import { ApiService } from './api.service';
import { TransactionDto } from 'src/common/dtos/transaction.dto';
import { DateInterval, getYearlyIntervalFromDate } from './date-utils';
import { AccountDto } from 'src/common/dtos/account.dto';

@Injectable()
export class AccountsService {
  constructor(private readonly apiService: ApiService) {}

  public async getUserAccounts(): Promise<AccountDto[]> {
    return await this.apiService.getAccounts();
  }

  public async getAllTransactions(
    accounts: AccountDto[],
  ): Promise<TransactionDto[]> {
    const allTransactionsForEachAccount = accounts.map(
      async (account) =>
        await this.getAllTransactionsFromAccount(account.account_id),
    );

    return this.flattenArrayOfPromiseOfArray(allTransactionsForEachAccount);
  }

  public async getAllTransactionsFromAccount(
    accountId: string,
  ): Promise<TransactionDto[]> {
    const oldestTransaction = await this.apiService.getOldestTransaction(
      accountId,
    );
    const dateIntervals = getYearlyIntervalFromDate(
      new Date(oldestTransaction?.timestamp),
    );

    const transactionsPerDateInterval = dateIntervals.map(
      async (interval: DateInterval) =>
        this.apiService.getTransactions(
          accountId,
          interval.start,
          interval.end,
        ),
    );

    return this.flattenArrayOfPromiseOfArray(transactionsPerDateInterval);
  }

  private async flattenArrayOfPromiseOfArray<T>(
    array: Promise<T[]>[],
  ): Promise<T[]> {
    const arrayOfArray = await Promise.all(array);
    return arrayOfArray.flat(1);
  }
}
