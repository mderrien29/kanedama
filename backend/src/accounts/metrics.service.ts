import { Injectable } from '@nestjs/common';

import { TransactionDto } from 'src/common/dtos/transaction.dto';
import { AccountDto } from 'src/common/dtos/account.dto';
import { AnswerDto } from 'src/common/dtos/answer.dto';
import { substractMonthToDate, substractYearsToDate } from './date-utils';

@Injectable()
export class MetricsService {
  public getUserMetrics(
    accounts: AccountDto[],
    transactions: TransactionDto[],
  ): AnswerDto {
    const transactionsOrderedByDate = this.sortTransactionsByDate(transactions);
    const accountsBalanceOverTime = this.calculateAccountAmountOverTime(
      transactionsOrderedByDate,
      accounts,
    );
    const { min, max } = this.calculateMinMaxBalance(accountsBalanceOverTime);

    const hadRecentActivity = this.hadRecentActivityLastYears(
      transactionsOrderedByDate,
      3,
    );
    const averageRecentIncome = this.calculateAverageIncome(transactions, 6);

    return this.formatUserMetrics(
      min,
      max,
      averageRecentIncome,
      hadRecentActivity,
    );
  }

  private calculateAccountAmountOverTime(
    transactionsOrderedByDate: TransactionDto[],
    accounts: AccountDto[],
  ): number[] {
    const currentAccountValue = this.sumAccountsCurrentAmount(accounts);
    const amountOverTime = [currentAccountValue];

    for (let i = 0; i < transactionsOrderedByDate.length; i++) {
      amountOverTime[i + 1] =
        amountOverTime[i] -
        transactionsOrderedByDate[transactionsOrderedByDate.length - i - 1]
          .amount;
    }

    return amountOverTime;
  }

  private calculateMinMaxBalance(amountOverTime: number[]): {
    min: number;
    max: number;
  } {
    // faster than using spread operator on large arrays
    const max = Math.max.apply(null, amountOverTime);
    const min = Math.min.apply(null, amountOverTime);
    return { min, max };
  }

  private hadRecentActivityLastYears(
    transactionsOrderedByDate: TransactionDto[],
    numberOfYears: number,
  ): boolean {
    const lastTransaction =
      transactionsOrderedByDate[transactionsOrderedByDate.length - 1];
    const dateRecentTreshold = substractYearsToDate(numberOfYears);

    return this.isTransactionMoreRecentThan(
      lastTransaction,
      dateRecentTreshold,
    );
  }

  private calculateAverageIncome(
    transactionsOrderedByDate: TransactionDto[],
    numberOfMonths: number,
  ): number {
    const dateLastTransaction = new Date(
      transactionsOrderedByDate[transactionsOrderedByDate.length - 1].timestamp,
    );
    const dateRecentTreshold = substractMonthToDate(
      numberOfMonths,
      dateLastTransaction,
    );

    const positiveRecentTransactions = this.getPositiveRecentTransactions(
      transactionsOrderedByDate,
      dateRecentTreshold,
    );
    const totalIncome = this.sumTransactionsAmounts(positiveRecentTransactions);
    const averageIncome = totalIncome / positiveRecentTransactions.length;
    return averageIncome;
  }

  private getPositiveRecentTransactions(
    transactions: TransactionDto[],
    recentTreshold: Date,
  ): TransactionDto[] {
    return transactions.filter((transaction) => {
      return (
        this.isTransactionMoreRecentThan(transaction, recentTreshold) &&
        this.isTransactionPositive(transaction)
      );
    });
  }

  private formatUserMetrics(
    minBalance: number,
    maxBalance: number,
    averageIncomeLast6Month: number,
    activeLast3Years: boolean,
  ): AnswerDto {
    return {
      min_balance: Math.floor(minBalance),
      max_balance: Math.floor(maxBalance),
      '6_month_average_income': Math.floor(averageIncomeLast6Month),
      '3_years_activity': activeLast3Years,
    };
  }

  private sortTransactionsByDate(
    transactions: TransactionDto[],
  ): TransactionDto[] {
    return transactions.sort(this.compareTransactionDate.bind(this));
  }

  private compareTransactionDate(a: TransactionDto, b: TransactionDto): number {
    return new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime();
  }

  private isTransactionMoreRecentThan(
    transaction: TransactionDto,
    recentTreshold: Date,
  ): boolean {
    const transactionDate = new Date(transaction.timestamp);
    return transactionDate.getTime() > recentTreshold.getTime();
  }

  private isTransactionPositive(transaction: TransactionDto): boolean {
    return transaction.amount > 0;
  }

  private sumTransactionsAmounts(transactions: TransactionDto[]): number {
    return transactions.reduce(
      (sum, transaction) => (sum += transaction.amount),
      0,
    );
  }

  private sumAccountsCurrentAmount(accounts: AccountDto[]): number {
    return accounts.reduce((sum, account) => (sum += account.current), 0);
  }
}
