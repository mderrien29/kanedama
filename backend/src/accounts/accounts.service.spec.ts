import { Test, TestingModule } from '@nestjs/testing';

import { AccountsService } from './accounts.service';

describe('AccountsService', () => {
  let accountsService: AccountsService;

  beforeEach(async () => {
    const app: TestingModule = await Test.createTestingModule({
      providers: [AccountsService],
    }).compile();

    accountsService = app.get<AccountsService>(AccountsService);
  });

  describe.skip('getUserAccounts', () => {});
  describe.skip('getAllTransactions', () => {});
  describe.skip('getAllTransactionsFromAccount', () => {});
});
