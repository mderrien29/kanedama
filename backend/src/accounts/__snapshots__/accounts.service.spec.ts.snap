// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountsService getAllTransactions should return the expected result 1`] = `
Array [
  Object {
    "amount": 1000,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2019-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 10,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -20,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-31",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 5,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-10-20",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-11-01",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 1000,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2019-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 10,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -20,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-31",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 5,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-10-20",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-11-01",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 1000,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2019-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 10,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -20,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-31",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 5,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-10-20",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-11-01",
    "transaction_category": "category",
    "transaction_type": "type",
  },
]
`;

exports[`AccountsService getAllTransactionsFromAccount should call ApiService.getOldestTransaction with the expected parameters 1`] = `
Array [
  Array [
    "accountId",
  ],
]
`;

exports[`AccountsService getAllTransactionsFromAccount should call ApiService.getTransactions with the expected parameters 1`] = `
Array [
  Array [
    "accountId",
    2019-08-30T00:00:00.000Z,
    2020-08-29T00:00:00.000Z,
  ],
  Array [
    "accountId",
    2020-08-29T00:00:00.000Z,
    2021-08-29T00:00:00.000Z,
  ],
  Array [
    "accountId",
    2021-08-29T00:00:00.000Z,
    2021-12-01T00:00:00.000Z,
  ],
]
`;

exports[`AccountsService getAllTransactionsFromAccount should return the expected result 1`] = `
Array [
  Object {
    "amount": 1000,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2019-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 10,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -20,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-31",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 5,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-10-20",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-11-01",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 1000,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2019-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 10,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -20,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-31",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 5,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-10-20",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-11-01",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 1000,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2019-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 10,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -20,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-30",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-08-31",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": 5,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-10-20",
    "transaction_category": "category",
    "transaction_type": "type",
  },
  Object {
    "amount": -100,
    "currency": "GBP",
    "status": "status",
    "timestamp": "2020-11-01",
    "transaction_category": "category",
    "transaction_type": "type",
  },
]
`;

exports[`AccountsService getUserAccounts should return the expected result 1`] = `
Array [
  Object {
    "account_id": "id",
    "account_number": "1234",
    "account_type": "ACCOUNT",
    "available": 10000,
    "currency": "USD",
    "current": 10000,
    "iban": "iban",
    "sort_code": "sort",
    "swift_bic": "bic",
    "update_timestamp": "2021-08-30",
  },
]
`;
